storage/registers.go

```diff
@@ -0,0 +1,43 @@
+package storage
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// RegisterIndex defines methods for the register index.
+type RegisterIndex interface {
+	RegisterIndexReader
+	RegisterIndexWriter
+}
+
+// RegisterIndexReader defines read-only operations on the register index.
+type RegisterIndexReader interface {
+	// Get register by the register ID at a given block height.
+	//
+	// If the register at the given height was not indexed, returns the highest
+	// height the register was indexed at.
+	// An error is returned if the register was not indexed at all or if the height is out of bounds.
+	// Expected errors:
+	// - storage.ErrNotFound if the register was not found in the db or is out of bounds.
+	Get(ID flow.RegisterID, height uint64) (flow.RegisterValue, error)
+	// LatestHeight returns the latest indexed height.
+	// Expected errors:
+	// - storage.ErrNotFound if no heights have been indexed. This should only happen when the database is first initialized.
```

Comment:
This won't happen, because when database is first initialized, it will be initialized as the root block height, same as for `FirstHeight`.

Reply:
There is an edge case for this: 
- The checkpoint file is being ingested
- values for root block is queried

so the `storage.ErrHeightNotIndexed` should still be returned in this case

Reply:
>The checkpoint file is being ingested

Do you mean the checkpoint file is being imported to pebble?

We would not start the service until the checkpoint file importing is done, which means, if we ever start the service, the checkpoint has have been imported, and the root height should be ready.

The root height is from the root snapshot, always available

Reply:
Ah okay understood, will update comment above then, the edge case only applies if you startup `index`

Reply:
> This won't happen, because when database is first initialized, it will be initialized as the root block height, same as for FirstHeight.

@zhangchiqing Shouldn't `FirstHeight` be the first height indexed, not the node's root height? The uninitialized db has no way to know which will be the first height until the first register is actually indexed.